myhoard/web_server.py

```diff
@@ -78,6 +80,42 @@ async def backup_list(self, _request):
                     response["backups"] = self.controller.state["backups"]
                 return json_response(response)
 
+    async def backup_preserve(self, request):
+        with self._handle_request(name="backup_preserve"):
+            body = await self._get_request_json(request)
+            stream_id = body.get("stream_id")
+            preserve_until = body.get("preserve_until")
+
+            if stream_id is None or preserve_until is None:
+                raise BadRequest("`stream_id` and `preserve_until` must be set in request body.")
+
+            try:
+                preserve_until = datetime.fromisoformat(preserve_until).astimezone(timezone.utc)
+                now = datetime.now(timezone.utc)
+                if preserve_until < now:
+                    raise BadRequest("`preserve_until` must be a date in the future.")
+            except ValueError:
+                raise BadRequest("`preserve_until` must be a valid isoformat datetime string.")
+
+            return self._handle_preservation_request(PreservationRequest(stream_id=stream_id, preserve_until=preserve_until))
+
+    async def backup_cancel_preserve(self, request):
+        with self._handle_request(name="backup_cancel_preserve"):
+            body = await self._get_request_json(request)
+            stream_id = body.get("stream_id")
+            if stream_id is None:
+                raise BadRequest("`stream_id` must be set in request body.")
+
+            return self._handle_preservation_request(PreservationRequest(stream_id=stream_id, preserve_until=None))
+
+    def _handle_preservation_request(self, preservation_request: PreservationRequest):
+        try:
+            with self.controller.lock:
+                self.controller.mark_backup_preservation(preservation_request)
+                return json_response({"success": True})
+        except Exception as ex:
+            raise BadRequest(str(ex))
```

Comment:
Did `PUT /backup/{stream_id}/preserve` instead. And now it returns 500 in case there was an internal error (something wrong when writing the file for example)